File: backend/app/schemas/ar.py
from typing import List, Dict, Any, Optional
from pydantic import BaseModel, Field
from datetime import datetime

class ARPointBase(BaseModel):
    """Base schema for AR point data"""
    title: str
    description: str
    latitude: float
    longitude: float
    altitude: Optional[float] = None
    type: str
    
class ARPointCreate(ARPointBase):
    """Schema for creating a new AR point"""
    metadata: Dict[str, Any] = Field(default_factory=dict)
    
class ARPoint(ARPointBase):
    """Schema for AR point response"""
    id: str
    metadata: Dict[str, Any]
    
    class Config:
        orm_mode = True
        
class HistoricalARPointCreate(ARPointCreate):
    """Schema for creating a historical AR point"""
    year: int
    historical_context: str
    image_url: Optional[str] = None
    
class HistoricalARPoint(ARPoint):
    """Schema for historical AR point response"""
    year: int
    historical_context: str
    image_url: Optional[str] = None
    
class NavigationARPointCreate(ARPointCreate):
    """Schema for creating a navigation AR point"""
    distance: float
    eta: Optional[int] = None
    direction: str
    
class NavigationARPoint(ARPoint):
    """Schema for navigation AR point response"""
    distance: float
    eta: Optional[int] = None
    direction: str
    
class NatureARPointCreate(ARPointCreate):
    """Schema for creating a nature AR point"""
    species: Optional[str] = None
    ecosystem_info: Optional[str] = None
    conservation_status: Optional[str] = None
    
class NatureARPoint(ARPoint):
    """Schema for nature AR point response"""
    species: Optional[str] = None
    ecosystem_info: Optional[str] = None
    conservation_status: Optional[str] = None
    
class ARPointRequest(BaseModel):
    """Schema for requesting AR points"""
    latitude: float
    longitude: float
    radius: Optional[float] = 500
    types: Optional[List[str]] = None
    
class ARRenderSettingsUpdate(BaseModel):
    """Schema for updating AR render settings"""
    distance_scale: Optional[float] = None
    opacity: Optional[float] = None
    color_scheme: Optional[str] = None
    show_labels: Optional[bool] = None
    show_distances: Optional[bool] = None
    show_arrows: Optional[bool] = None
    animation_speed: Optional[float] = None
    detail_level: Optional[int] = None
    accessibility_mode: Optional[bool] = None
    
class HistoricalOverlayRequest(BaseModel):
    """Schema for requesting a historical overlay"""
    latitude: float
    longitude: float
    year: Optional[int] = None
    
class HistoricalOverlayResponse(BaseModel):
    """Schema for historical overlay response"""
    title: str
    year: int
    description: str
    key_features: List[str]
    daily_life: str
    image_url: Optional[str] = None
    latitude: float
    longitude: float
    
class ARViewParameters(BaseModel):
    """Schema for AR view parameters"""
    device_heading: float
    device_pitch: float
    camera_fov: Optional[float] = 60.0
    
class RenderableARElement(BaseModel):
    """Schema for renderable AR element"""
    id: str
    source_point_id: str
    view_x: float
    view_y: float
    view_z: float
    scale: float
    opacity: float
    visible: bool
    appearance: Dict[str, Any]
    interaction: Dict[str, Any]
    
class ARRenderResponse(BaseModel):
    """Schema for AR render response"""
    elements: List[RenderableARElement]
    settings: Dict[str, Any]
    timestamp: datetime = Field(default_factory=datetime.now)

File: mobile/src/types/ar.ts
import { Timestamp } from './global.d';

// Basic AR point types

export interface ARPointBase {
  id: string;
  title: string;
  description: string;
  latitude: number;
  longitude: number;
  altitude?: number;
  type: string;
  metadata: Record<string, any>;
}

export interface HistoricalARPoint extends ARPointBase {
  year: number;
  historical_context: string;
  image_url?: string;
}

export interface NavigationARPoint extends ARPointBase {
  distance: number;
  eta?: number;
  direction: string;
}

export interface NatureARPoint extends ARPointBase {
  species?: string;
  ecosystem_info?: string;
  conservation_status?: string;
}

// Combine the point types into a discriminated union
export type ARPointResponse = 
  | (ARPointBase & { type: 'generic' }) 
  | (HistoricalARPoint & { type: 'historical' })
  | (NavigationARPoint & { type: 'navigation' })
  | (NatureARPoint & { type: 'nature' });

// Request/response types for API endpoints

export interface ARPointRequest {
  latitude: number;
  longitude: number;
  radius?: number;
  types?: string[];
}

export interface ARViewParameters {
  device_heading: number;
  device_pitch: number;
  camera_fov?: number;
}

export interface RenderableARElement {
  id: string;
  source_point_id: string;
  view_x: number;
  view_y: number;
  view_z: number;
  scale: number;
  opacity: number;
  visible: boolean;
  appearance: {
    icon?: string;
    color?: string;
    highlight_color?: string;
    show_year?: boolean;
    show_image?: boolean;
    show_distance?: boolean;
    show_eta?: boolean;
    show_species?: boolean;
    show_labels?: boolean;
    frame_style?: string;
    text_style?: string;
    animation?: string;
    arrow_type?: string;
    pulse_effect?: boolean;
    year?: number | string;
    distance?: number | string;
    eta?: number | string;
    species?: string;
    ecosystem_info?: string;
    conservation_status?: string;
    [key: string]: any;
  };
  interaction: {
    tappable?: boolean;
    expanded_view?: boolean;
    audio_feedback?: boolean;
    haptic_feedback?: boolean;
    show_timeline?: boolean;
    show_full_description?: boolean;
    show_gallery?: boolean;
    comparison_view?: boolean;
    show_full_instructions?: boolean;
    path_preview?: boolean;
    recenter_map?: boolean;
    estimate_time?: boolean;
    show_species_info?: boolean;
    show_ecosystem_info?: boolean;
    show_conservation_status?: boolean;
    camera_integration?: boolean;
    [key: string]: any;
  };
  title?: string;
  description?: string;
}

export interface ARRenderResponse {
  elements: RenderableARElement[];
  settings: ARRenderSettings;
  timestamp: string;
}

export interface ARRenderSettings {
  distance_scale: number;
  opacity: number;
  color_scheme: string;
  show_labels: boolean;
  show_distances: boolean;
  show_arrows: boolean;
  animation_speed: number;
  detail_level: number;
  accessibility_mode: boolean;
}

export interface ARRenderSettingsUpdate {
  distance_scale?: number;
  opacity?: number;
  color_scheme?: string;
  show_labels?: boolean;
  show_distances?: boolean;
  show_arrows?: boolean;
  animation_speed?: number;
  detail_level?: number;
  accessibility_mode?: boolean;
}

export interface HistoricalOverlayRequest {
  latitude: number;
  longitude: number;
  year?: number;
}

export interface HistoricalOverlayResponse {
  title: string;
  year: number;
  description: string;
  key_features: string[];
  daily_life: string;
  image_url?: string;
  latitude: number;
  longitude: number;
}